trapi_throttle/server.py

```diff
@@ -25,16 +39,154 @@
+
+    # Subscribe to changes to the buffer
+    kp_buffer_pattern = f"__keyspace@0__:{kp_id}:buffer:*"
+    channel, = await conn.psubscribe(kp_buffer_pattern)
+
+    # TODO figure out why we need to sleep here
+    # for our tests to pass
+    await asyncio.sleep(1)
+
+    # Wait for anything to be added to the buffer
+    while await channel.wait_message():
+        await channel.get()
+
+        # Check if we actually have work to do
+        batch_keys = await APP.state.redis.keys(f"{kp_id}:buffer:*")
+        if len(batch_keys) == 0:
+            continue
+
+        LOGGER.debug(
+            f"Processing batch of size {len(batch_keys)} for KP {kp_id}"
+        )
+
+        now = datetime.datetime.utcnow()
+        tat = datetime.datetime.fromisoformat(await tat_db.get())
+
+        time_remaining_seconds = (tat - now).total_seconds()
+        # Wait for TAT
+        if time_remaining_seconds > 0:
+            LOGGER.debug(f"Waiting {time_remaining_seconds}")
+            await asyncio.sleep(time_remaining_seconds)
+
+        # Process batch
+        batch_request_ids = [key.split(':')[-1] for key in batch_keys]
+
+        request_values_db = [
+            RedisValue(APP.state.redis, f"{kp_id}:buffer:{request_id}")
+            for request_id in batch_request_ids
+        ]
+        request_values = await asyncio.gather(*[
+            request_value_db.get() for request_value_db in request_values_db
+        ])
+
+        # TODO Implement merging utility
+        merged_request_value = request_values[0]
+
+        async with httpx.AsyncClient() as client:
+            response = await client.post(kp_info.url, json = merged_request_value)
+
+        # TODO Implement split utility
+        for response_id in batch_request_ids:
+            await APP.state.redis.set(
+                f"{kp_id}:finished:{response_id}",
+                response.content,
+            )
+
+        # Update TAT
+        interval = kp_info.request_duration / kp_info.request_qty
+        new_tat = datetime.datetime.utcnow() + interval
+        await tat_db.set(new_tat.isoformat())
+
+    # Close redis connection
+    conn.close()
+    await conn.wait_closed()
+
+@APP.post("/register/{kp_id}")
+async def register_kp(
+        kp_id: str,
+        kp_info: KPInformation,
+        request: Request,
+):
+    kp_info_db = RedisValue(request.app.state.redis, f"{kp_id}:info")
+    await kp_info_db.set(kp_info.json())
+
+    loop = asyncio.get_event_loop()
+    loop.create_task(process_batch(kp_id))
+
+    return {"status" : "created"}
+
+
 @APP.post('/query/{kp_id}')
-async def async_query(
+async def query(
         kp_id: str,
         query: Query,
 ) -> Query:
-    """Queue up a query for batching and return when completed"""
-    return query
+    """ Queue up a query for batching and return when completed """
+
+
+    # Insert query into db for processing
+    request_id = uuid.uuid1()
+    query_input_db = RedisValue(APP.state.redis, f"{kp_id}:buffer:{request_id}")
```

Comment:
Is there a reason to use shared connections here if we're going to create a new connection in 2 lines?

Reply:
The shared connection is a connection pool while the new connection is just a single one. I think that this approach where we use the connection pool for all data access and new connections for subscriptions makes the most sense.

Reply:
Why not use the subscription connection for data access as well?

Reply:
What's the motivation for segregating the connections' responsibilities?

Reply:
Ah sorry I missed this. When you subscribe to a channel it alters the connection so it is not possible to run any redis commands other than unsubscribe after a subscription. This can cause problems with a connection pool because the underlying connection might be put in subscribe mode and then handed to a different part of the code where it won't work.

The subscription doesn't actually return any values. It only returns keys so we need to fetch the values separately.